v1.9 backports 2021-03-02

jrfastab opened the pull request:
* #14952 -- Add configurable suffix for operator image repo in helm (@nebril)
 * #14999 -- cilium: encryption fix, ipv4-pod-subnets without encryptnode fails (@jrfastab)
 * #15039 -- ipsec: Use 64bits for XFRM output sequence number (@pchaigno)
 * #15048 -- cilium: encryption, fixes for ENI & Azure mode with shared podIPs and networkIPs (@jrfastab)
 * #15058 -- ci: disable endpointslice in 1.17 cluster (@nebril)
 * #15081 -- ci: use quay.io images in upstream tests (@nebril)
 * #15107 -- pkg/kvstore: fix etcd segmentation violation (@aanm)
 * #15096 -- checkpatch: update image (skip backports, add a check, suppress one report type) (@qmonnet)
 * #15118 -- backporting: Update instructions for backporting workflow (@aditighag)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 14952 14999 15039 15048 15058 15081 15107 15096 15118; do contrib/backporting/set-labels.py $pr done 1.9; done
```

nebril committed:
helm: Add suffix for operator image

[ upstream commit f0d37f8e9a17b579401b8c002966bb731aff5d74 ]

Because operator has cloud provider custom suffixes we are not able to add custom
suffixes (like `-dev` or `-ci`) for the operator image seamlesly.

New operator-specific option allows to do this.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

nebril committed:
test: add `ciliu.operator-suffix` cli option

[ upstream commit 114130883bed3ec906ff172604b1f8142547eefa ]

This option makes it possible to use new operator.image.suffix Helm
value in test runs.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrfastab committed:
cilium: encryption fix, ipv4-pod-subnets without encryptnode fails

[ upstream commit f93a2fdd70648f760e73bde8c0d3e6a494177512 ]

In the default mode encryption picks the outer IP headers to use
based on the destination IP of the packet to be encrypted. This
works because we create a IPSec rule that matches the destination
IP of the packet using the node CIDR IP used to allocate IPs for
that node. This allows us to scale rules with number of nodes
instead of number of pods.

However, in modes where a global IP pool is used for any pods we
no longer have a subnet -> pod mapping that we use above. To
handle this case, instead of adding a rule per pod, we rewrite
the srcIP,dstIP after encryption using the ipcache. The ipcache
has the pod->node mapping so we can use this reliable. Then we
do a fib lookup to rewrite the src/dst MAC and redirect out the
correct egress interface.

But, we have a bug where nodeEncrypt routing rules are added
even if encryptNode is not enabled. Additionally, when encryptNode
is enabled we add wrong and possibly conflicting rules from the
normal path.

To fix ensure we only add encryptNode routes correctlyy from the
subnetEncryption path with ipv4-pod-subnets is enabled.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

pchaigno committed:
ipsec: Use 64bits for XFRM output sequence number

[ upstream commit 4ea12ee32e48ea08de72aa36d45f8df0f5f468bf ]

The anti-replay mechanism of IPSec has several implementations in the
Linux kernel [1]. The legacy implementation, which we are currently
using, relies on a 32bit output sequence number. On cluster with a high
packet throughput and less frequent key rotations, this number can reach
its maximum value quickly. In such cases, the RFC states [2] that the
packet should be dropped and userspace should take care of replacing the
xfrm state entry.

To reduce the likelihood of that happening, we can use the ESN
implementation of the anti-replay mechanism, which relies on a 64bits
sequence number.

Regardless of the sequence number size, we currently rely on the user to
implement key management outside of Cilium. To prevent issues as
described above from happening in the longer term, we may want to handle
the key management ourselves.

1 - https://elixir.bootlin.com/linux/v5.11/source/net/xfrm/xfrm_replay.c#L707
2 - https://tools.ietf.org/html/rfc4303#page-25
Co-authored-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, allow programming IPSec FWD policy and template IPs

[ upstream commit d02b6b74447335abe36c6b9b59027c5417d647ee ]

So far we've programmed FWD policy rules to mirror the IN policy rules,
but this requires the outer IPSec tunnel dstIP and inner dstIP cidr to
be non-overlapping. Also we've been setting the template IPs the same
as the matching src/dst IPs. We are about to fix the ENI case where these
assumptions are no longer true.

In order to support this allow pushing unique FWD/IN policies and
also specify matching IPs and template IPs.

Note: It probably makes sense to do a heavier refactor here and rip
out UpsertIPsecEndpoint(), at this point it just adds confusion in
my opinion. But, lets avoid that at the moment for backports.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, set optional flag for fwd policy rules

[ upstream commit 53650a85a3f9b43c226449be045a966859a697f6 ]

The XFRM stack will check for fwd policies after a decrypt or encrypt
operation is done. We don't really need these for policy reasons
because Cilium BPF side is enforcing policy. We can't just delete
the rules though. If we did the decrypted traffic, which will still
have the xfrm context encoded in the skb, will fail policy match and
generate a NOPOLICY error. But as is in ENI modes the template lookup
is failing with an TMPL_MISMATCH error.

To fix make the fwd rules less strict by adding the optional bit
to the rule. This will skip the template lookup altogether on the
kernel side.

Although this is needed for ENI cases where the packet is handled
by the routing stack we can enable it in all cases. Making other
cases less strict is also OK.

Before this we had fwd `ip x p` like this,

root@ip-192-168-90-141:/home/cilium# ip x p
src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.0.0
		proto esp reqid 1 mode tunnel

Now we have a 'level use' included indicating the policy is not strict,

root@ip-192-168-90-141:/home/cilium# ip x p
src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.0.0
		proto esp reqid 1 mode tunnel
		level use

This corresponds to kernel code,

 static inline int
 xfrm_policy_ok(const struct xfrm_tmpl *tmpl, const struct sec_path *sp, int start,
	       unsigned short family)
 {
	int idx = start;

	if (tmpl->optional) {
		if (tmpl->mode == XFRM_MODE_TRANSPORT)
			return start;
	} else
		start = -1;
	for (; idx < sp->len; idx++) {
		if (xfrm_state_ok(tmpl, sp->xvec[idx], family))
			return ++idx;
		if (sp->xvec[idx]->props.mode != XFRM_MODE_TRANSPORT) {
			if (start == -1)
				start = -2-idx;
			break;
		}
	}
	return start;
 }

And with 'level use' we hit the first 'if (tmpl->optional)' so that the
subsequent xfrm_state_ok checks are skipped.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, ENI with subnet encryption needs rules for link IP

[ upstream commit 86d78df2650d7bc7b416746f025a93eb800bf45c ]

In ENI mode we use link IP for outer IPsec headers. In order for rules
to be configured with correct IPs do a link lookup to discover the
correct IP.

Without this patch we will be missing the 'in' policy rules shown here,

 root@ip-192-168-90-141:/home/cilium# ip x p
 src 0.0.0.0/0 dst 192.168.0.0/16
	dir fwd priority 0 ptype main
	tmpl src 0.0.0.0 dst 192.168.90.141
		proto esp reqid 1 mode tunnel
		level use
 src 0.0.0.0/0 dst 192.168.64.0/19
 	dir in priority 0 ptype main
	mark 0xd00/0xf00
	tmpl src 0.0.0.0 dst 192.168.90.141

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, fix redirect when endpoint routes enabled

[ upstream commit 287f49c293b31bc8044239fd29b27e829e798e80 ]

If we have endpoint routes enabled we want to use the routing table
to find the lxc* device. Trying to do this with redirects doesn't
work and may cause cilium_host to drop the packet.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, ENI mode uses overlapping CIDR for outer/inner IPs

[ upstream commit eb1ca8515eff62d7e7c151b1a7aa394c94153b80 ]

TLDR: Remove 0xd00 rule/routes and let xfrm stack clear mark, but restrict
      overlapping pod/network CIDRs to 4.19 and above kernels.

Because ENI mode uses an outer IP that overlaps with the inner pod IPs we
need to be extra careful with the routing/rule configurations. For example
we currently put a route in the decrypt routing table that looks like
this,

	192.168.0.0/16 dev cilium_host mtu 9001

where 192.168.0.0/16 is the subnet IP in this example. The purpose of this
rule is twofold. First in the encryption direction it should match the
encrypted packet and send it to cilium_host for necessary rewrites. Second
on decryption there should be a simliar rule to match and send traffic to
the sending interface. Something like this,

	$POD_CIDR dev eth0 mtu

But, in the ENI case the decrypted CIDR and the encrypted CIDR are the
same so they collide and we get a single rule. Worse now received ESP
packets will match above and be routed to cilium_host before they can
be decrypted. cilium_host will become confused and send them to the
stack. At this point the mark set by bpf_network will be dropped and
the stack will miss the xfrm state lookup.

  eth0 (rx esp) -> bpf_packet (mark for decryption mark=0xd00)
                     -> rule finds 0x0d00 uses routing table 200
		     -> route hits {$CIDR dev cilium_host}
		     -> ESP packet on cilium host runs bpf_host
		     -> bpf_host return CTX_ACT_OK
		     -> cilium_host fwd to cilium_net mark = 0 now
		     -> ESP packet (dev=cilium_net) lookup xfrm state
		     -> XFRM_NOSTATE error because mark cleared

So everything went sideways after the 0xd00 routing table hit. The reason
we even have a 0xd00 routing rule is to clear the mark bits after decryption,
but before pushing to stack. Expected flow,

   eth0 (rx esp) -> bpf_packet (mark for decryption mark=0xd00)
                     -> rule finds 0x0d00 uses routing table 200
		     -> no match send to stack with mark=0xd00 dev=eth0
		     -> xfrm state lookup matches
		     -> decrypted packet mark=0xd00
                        rule for 0x0d00 uses tbl 200
		     -> route decrypted packet to eth0
 		     -> bpf_packet runs clears mark and sends to stack

At this point a decrypted packet is sent to stack just as if it had arrived
decrypted. We could remove the steps after the xfrm decrypt if the xfrm stack
cleared the mark with an output_mark action in xfrm. We haven't done this yet
because it doesn't work pre-4.19 kernels. For ENI + overlapping subnets we
will put the 4.19 kernel requirement and remove the 0xd00 rule completely.
This simplifies datapath to just,

  eth0 (rx esp) -> bpf_packet mark -> decrypt -> stack ...

Much cleaner, but only usable where the output mark can be set. So we
keep the fallback path to use the mark if the mark action is not
available.

Before this patch two ip rules will be set for encrypt and decrypt,

 $ ip rule
   1:      from all fwmark 0xe00/0xf00 lookup 200
   1:      from all fwmark 0xd00/0xf00 lookup 200
   ...

After this patch only the encrypt rule will be set,

 $ ip rule
   1:      from all fwmark 0xe00/0xf00 lookup 200
   ...

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, remove mark from fwd rule

[ upstream commit 031455357e0540f9c83398bc8593ee8284a5f0b3 ]

Make the fwd policy rule more permissive and remove the mark. This
way it will match in both cases: the mark is set because we are
running without subnet set, the case where the mark is 0 in the
ENI case.

Before this patch an example fwd policy was,

src 0.0.0.0/32 dst 192.168.32.0/19
        dir fwd priority 0 ptype main
        mark 0xd00/0xf00
        tmpl src 0.0.0.0 dst 192.168.62.241
                proto esp reqid 1 mode tunnel
                level use

Notice the 'mark 0xd00/0xf00' case restricting the match. After thiis
patch we drop the mark, which will make the match more permissive.

src 0.0.0.0/32 dst 192.168.32.0/19
        dir fwd priority 0 ptype main
        tmpl src 0.0.0.0 dst 192.168.62.241
                proto esp reqid 1 mode tunnel
                level use

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, clear mark from xfrm state in endpoint route modes

[ upstream commit bc8bb3c97b9217657a3b592e1259e3d7f8fa5f9f ]

In endpoint route modes we have routes in the system to the endpoints so we
can skip the route back to the interface with bpf_network.o set, previously
used to do the redirect needed when the stack can not route.

This will work back to 4.14 kernels where XFRM_OUTPUT_MARK was added. If we
need to go back further an alternative route/mark/rule will need to be
added.

Before this patch we had in state rules like this,

src 0.0.0.0 dst 192.168.62.241
        proto esp spi 0x00000003 reqid 1 mode tunnel
        replay-window 0
        mark 0xd00/0xf00 output-mark 0xd00/0xf00

The above will use the 'output-mark' field to set the skb->mark after
decryption to 0xd00. But, after patch (and with endpoint routes enables)
we set the output-mark to 0 with the following state rule.

src 0.0.0.0 dst 192.168.62.241
        proto esp spi 0x00000003 reqid 1 mode tunnel
        replay-window 0
        mark 0xd00/0xf00 output-mark 0x0/0xf00

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: encryption, do not try to auto-discover if user provides iface

[ upstream commit 428b0b0dededab19efe95403381c7138bf22e57c ]

We currently ignore the user provided interface in ENI modes. Lets not
overwrite user input if they specify it.

Fixes: 2c5e192676 ("encryption/loader: Attach bpf_network program to correct interface for ENI mode")
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

nebril committed:
ci: disable endpointslice in 1.17 cluster

[ upstream commit 3fb2c6ace5c5bc8fcfd02b6cdc127216bf40f3db ]

This change makes sure that we don't enable EndpointSlice in 1.17
clusters, so we can make sure that Cilium picks up the Endpoint objects
created in tests.

This change is fixing CI on 1.17 clusters. The EndpointSlice->Endpoint
transition was made to fix GKE runs, which started running on 1.17
clusters. This ended up breaking 1.17 runs on non-gke clusters. This
change will fix 1.17 runs.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: vendor, use cilium/netlink until upstream patches land

[ upstream commit 49fe92ad035fe681af83c1eab95a52f0afb822a9 ]

For encryption use cases we need the optional parameter in Xfrm
template. While we wait for those patches to land upstream point
at cilium/netlink with included patches.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

nebril committed:
ci: use quay.io images in upstream tests

[ upstream commit 676c2add9849bd9ec00448c7d0b13c2ed6d73736 ]

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

qmonnet committed:
checkpatch: update image (skip backports, +1 check, -1 warning)

[ upstream commit 42fcb077ca8a5ce92e8d11cb591f240d33557e8c ]

Update the checkpatch image tag (and switch to quay.io) to benefit from
the following changes:

- Skip checkpatch validation on backport branches
- Suppress reports of type COMMIT_LOG_LONG_LINE
- Add a repo-wide check on commit subject length
- Use $GITHUB_REPOSITORY variable when retrieving commits

For skipping backports we could instead update the conditions for
running the job in the current repo, but that would be troublesome if we
want to mark the action as required in the future. Therefore, we
delegate the decision to the shell script in the checkpatch image: it
returns early if a PR is not based on the 'master' branch.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

aanm committed:
pkg/kvstore: fix etcd segmentation violation

[ upstream commit c6610045b599b75e396c5420ee3e780c0bc78cd3 ]

Fixes the following panic:

```
goroutine 531 [running]:
go.etcd.io/etcd/clientv3/concurrency.(*Session).Orphan(...)
	/go/src/github.com/cilium/cilium/vendor/go.etcd.io/etcd/clientv3/concurrency/session.go:90
go.etcd.io/etcd/clientv3/concurrency.(*Session).Close(0xc0017b2b70, 0xc003a40570, 0x2)
	/go/src/github.com/cilium/cilium/vendor/go.etcd.io/etcd/clientv3/concurrency/session.go:96 +0x2a
github.com/cilium/cilium/pkg/kvstore.(*etcdClient).Close(0xc001cfe000)
	/go/src/github.com/cilium/cilium/pkg/kvstore/etcd.go:1654 +0x48e
github.com/cilium/cilium/pkg/clustermesh.(*remoteCluster).restartRemoteConnection.func1(0x2b31500, 0xc000fed2c0, 0x3c79c80, 0x0)
	/go/src/github.com/cilium/cilium/pkg/clustermesh/remote_cluster.go:178 +0x40c
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc001933200)
	/go/src/github.com/cilium/cilium/pkg/controller/controller.go:205 +0xa77
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/go/src/github.com/cilium/cilium/pkg/controller/manager.go:120 +0xb50
```

Fixes: c1b05dfba81a ("pkg/kvstore: introduced a dedicated session for locks")
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

aditighag committed:
backporting: Update instructions for backporting workflow

[ upstream commit 198fe64d79b53c983ba74f09f5c7a1c34408666d ]

Commit 02320e added support for working with Cilium forks. Document the
additional steps that are required with this new workflow.

Additionally, workflow scope might be needed with the GitHub token in
some cases that update the corresponding yamls -

``(refusing to allow a Personal Access Token to create or update workflow `.github/workflows/go-check.yaml` without `workflow` scope``

Signed-off-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: John Fastabend <john.fastabend@gmail.com>

aanm reviewed (approved):
LGTM for my commits

pchaigno reviewed (approved):
:+1: for my PR.

qmonnet reviewed (approved):
> * #15096 -- checkpatch: update image (skip backports, add a check, suppress one report type) (@qmonnet)

Looks good to me, thanks!

nebril committed:
ci: remove params from upstream k8s job

[ upstream commit 637a0258aea1f50872ff3edb0f419dda02b3bfb4 ]

These params are not compatible with our master job configuration, which
causes master builds to fail.

These params are stored in jenkins PR upstream k8s job config, so there
is no need for them to be in jenkinsfile.

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Nicolas Busseneau <nicolas@isovalent.com>
